File: poly_poly_bot/polymarket/src/data-api-source.ts
// Data API polling source — wraps existing fetchAllTraderActivities with source interface

import { TradeSource } from "./trade-source";
import { fetchAllTraderActivities } from "./trade-monitor";
import { isSeenTrade, isMaxRetries } from "./trade-store";
import { enqueueTrade } from "./trade-queue";
import { isMarketEnded } from "./market-cache";
import { CONFIG } from "./config";
import { logger } from "./logger";
import { errorMessage } from "./types";
import { sleep } from "./utils";
import { telegram } from "./telegram-notifier";

export class DataApiSource implements TradeSource {
  name = "data-api";
  private running = false;

  async start(): Promise<void> {
    this.running = true;
    const CIRCUIT_BREAKER_THRESHOLD = 10;
    let consecutiveFailures = 0;
    let circuitBreakerAlerted = false;

    while (this.running) {
      try {
        const trades = await fetchAllTraderActivities();
        const now = Date.now();
        const maxAgeMs = CONFIG.maxTradeAgeHours * 60 * 60 * 1000;
        const newTrades = trades.filter(t => {
          if (isSeenTrade(t.id) || isMaxRetries(t.id)) return false;
          // Trade age gate: skip trades older than MAX_TRADE_AGE_HOURS.
          // Defense-in-depth against cursor resets after bot restart causing
          // old trades to be re-fetched from the Data API.
          const tradeTime = new Date(t.timestamp).getTime();
          if (isNaN(tradeTime) || now - tradeTime > maxAgeMs) return false;
          return true;
        });
        for (const trade of newTrades) {
          // Skip trades on markets that have already resolved/ended
          if (trade.conditionId) {
            try {
              if (await isMarketEnded(trade.conditionId)) {
                logger.debug(`Skipping trade on ended market: "${trade.market}"`);
                continue;
              }
            } catch { /* non-critical — proceed with enqueue */ }
          }
          enqueueTrade(trade, new Date(trade.timestamp).getTime(), "data-api");
        }
        consecutiveFailures = 0;
        circuitBreakerAlerted = false;
      } catch (err: unknown) {
        consecutiveFailures++;
        logger.error(`Data API error (${consecutiveFailures}x): ${errorMessage(err)}`);
        if (consecutiveFailures >= CIRCUIT_BREAKER_THRESHOLD && !circuitBreakerAlerted) {
          circuitBreakerAlerted = true;
          telegram.botError(`Data API circuit breaker: ${consecutiveFailures} failures. Last: ${errorMessage(err)}`);
        }
      }

      const backoff = consecutiveFailures > 0
        ? Math.min(CONFIG.fetchInterval * Math.pow(2, Math.min(consecutiveFailures, 6)), 300_000)
        : CONFIG.fetchInterval;
      await sleep(backoff);
    }
  }

  stop(): void {
    this.running = false;
  }
}

File: poly_poly_bot/src/copy_trading/data_api_source.py
"""Data API polling source -- wraps fetchAllTraderActivities with circuit breaker."""

import asyncio
import time
from src.copy_trading.trade_monitor import fetch_all_trader_activities
from src.copy_trading.trade_store import (
    is_seen_trade, is_max_retries, record_reaction_latency)
from src.copy_trading.trade_queue import enqueue_trade
from src.config import CONFIG
from src.logger import logger
from src.models import QueuedTrade
from src.utils import error_message


class DataApiSource:
    """Polls the Data API for trader activity with exponential backoff and circuit breaker."""

    name = "data-api"

    def __init__(self) -> None:
        self._running = False

    async def start(self) -> None:
        """Start polling loop with circuit breaker protection."""
        self._running = True
        circuit_breaker_threshold = 10
        consecutive_failures = 0
        circuit_breaker_alerted = False

        while self._running:
            try:
                trades = await fetch_all_trader_activities()
                new_trades = [
                    t for t in trades
                    if not is_seen_trade(t.id) and not is_max_retries(t.id)
                ]
                # One wall-clock read for the whole batch: every trade in it
                # was returned by the same poll, so they were all "seen" now.
                received_ms = time.time() * 1000
                for trade in new_trades:
                    from datetime import datetime
                    ts_ms = (
                        datetime.fromisoformat(
                            trade.timestamp.replace("Z", "+00:00")
                        ).timestamp()
                        * 1000
                    )
                    # ts_ms is the TARGET'S trade time; received_ms is when we
                    # saw it. The gap between them is the number that decides
                    # whether copying is fast enough to be worth real money,
                    # and until now it was recorded nowhere.
                    record_reaction_latency(received_ms - ts_ms)
                    # The api's clock on this fill; the chain stamps its own
                    # under the same id and the two-clocks report joins them.
                    try:
                        from src.copy_trading import two_clocks
                        two_clocks.note("data-api", trade.id, their_ts=ts_ms / 1000.0,
                                        seen_at=received_ms / 1000.0,
                                        target=trade.trader_address, token_id=trade.token_id)
                    except Exception as exc:  # measurement never blocks a copy
                        logger.warn(f"two-clocks stamp failed: {error_message(exc)}")
                    enqueue_trade(QueuedTrade(
                        trade=trade,
                        enqueued_at=ts_ms,
                        source_detected_at=ts_ms,
                        received_at_ms=received_ms,
                        source="data-api",
                    ))
                consecutive_failures = 0
                circuit_breaker_alerted = False
            except Exception as err:
                consecutive_failures += 1
                logger.error(
                    f"Data API error ({consecutive_failures}x): {error_message(err)}"
                )
                if (
                    consecutive_failures >= circuit_breaker_threshold
                    and not circuit_breaker_alerted
                ):
                    circuit_breaker_alerted = True
                    from src.copy_trading.telegram_notifier import telegram
                    await telegram.bot_error(
                        f"Data API circuit breaker: {consecutive_failures} failures. "
                        f"Last: {error_message(err)}"
                    )

            backoff = (
                min(
                    CONFIG.fetch_interval * (2 ** min(consecutive_failures, 6)),
                    300,
                )
                if consecutive_failures > 0
                else CONFIG.fetch_interval
            )
            await asyncio.sleep(backoff)

    def stop(self) -> None:
        """Signal the polling loop to stop."""
        self._running = False
